docs/essentials/gc/unmanaged/implementingdispose.md

```diff
@@ -1 +1,442 @@
+
+   [DllImport("kernel32.dll")]
+   private static extern int GetFileSize(SafeFileHandle hFile, out int lpFileSizeHigh);
+
+   // Define locals.
+   private bool disposed = false;
+   private SafeFileHandle safeHandle; 
+   private long bufferSize;
+   private int upperWord;
+
+   public DisposableStreamResource(string filename)
+   {
+      if (filename == null)
+         throw new ArgumentNullException("The filename cannot be null.");
+      else if (filename == "")
+         throw new ArgumentException("The filename cannot be an empty string.");
+
+      IntPtr handle = CreateFile(filename, GENERIC_READ, FILE_SHARE_READ,
+                                 IntPtr.Zero, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL,
+                                 IntPtr.Zero);
+      if (handle != INVALID_HANDLE_VALUE)
+         safeHandle = new SafeFileHandle(handle, true);
+      else
+         throw new FileNotFoundException(String.Format("Cannot open '{0}'", filename));
+
+      // Get file size.
+      bufferSize = GetFileSize(safeHandle, out upperWord); 
+      if (bufferSize == INVALID_FILE_SIZE)
+         bufferSize = -1;
+      else if (upperWord > 0) 
+         bufferSize = (((long)upperWord) << 32) + bufferSize;
+   }
+
+   public long Size 
+   { get { return bufferSize; } }
+
+   public void Dispose()
+   {
+      Dispose(true);
+      GC.SuppressFinalize(this);
+   }           
+
+   protected virtual void Dispose(bool disposing)
+   {
+      if (disposed) return;
+
+      // Dispose of managed resources here.
+      if (disposing)
+         safeHandle.Dispose();
+
+      // Dispose of any unmanaged resources not wrapped in safe handles.
+
+      disposed = true;
+   }  
+}
+```
+
+## Using a safe handle to implement the dispose pattern for a derived class
+
+The following example illustrates the dispose pattern for a derived class, `DisposableStreamResource2`, that inherits from the `DisposableStreamResource` class presented in the previous example. The class adds an additional method, `WriteFileInfo`, and uses a `SafeFileHandle` object to wrap the handle of the writable file.
+
+```csharp
+using Microsoft.Win32.SafeHandles;
+using System;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+public class DisposableStreamResource2 : DisposableStreamResource
+{
+   // Define additional constants.
+   protected const uint GENERIC_WRITE = 0x40000000; 
+   protected const uint OPEN_ALWAYS = 4;
+
+   // Define additional APIs.
+   [DllImport("kernel32.dll")]   
+   protected static extern bool WriteFile(
+                                SafeFileHandle safeHandle, string lpBuffer, 
+                                int nNumberOfBytesToWrite, out int lpNumberOfBytesWritten,
+                                IntPtr lpOverlapped);
+
+   // Define locals.
+   private bool disposed = false;
+   private string filename;
+   private bool created = false;
+   private SafeFileHandle safeHandle;
```

Comment:
I think `safeHandle` is too generic, a better name would be something like `fileInfoHandle`.